core/app/c/[communitySlug]/settings/tokens/PermissionField.tsx

```diff
@@ -123,19 +124,100 @@ type ScopesWithOnlyNormalConstraints<
 
 type CustomConstraintFormElement<Value> = (props: {
 	value: boolean | Value;
-	onChange: (...args: any[]) => void;
+	onChange: (value: boolean | Value) => void;
 }) => React.ReactNode;
 
-const CreateTokenFormContextContext = createContext<CreateTokenFormContext>({ stages: [] });
-
 /**
  * Here you configure the specific form elements for each permission type
  */
 const permissionContraintMap: PermissionContraintMap = {
 	community: null,
 	pub: {
+		[ApiAccessType.read]: ({ value, onChange }) => {
+			const context = useContext(CreateTokenFormContext);
+
+			return (
+				<div className="flex flex-col gap-2">
+					<h3 className="font-semibold">Stages</h3>
+					<span className="text-xs text-muted-foreground">
+						Select the stages this token can read Pubs from
+					</span>
+					<MultiSelect
+						variant="inverted"
+						options={context.stages.allOptions}
+						defaultValue={
+							value === true || !value ? context.stages.allValues : value.stages
+						}
+						onValueChange={(val) => {
+							const allStagesSelected =
+								val.length === context.stages.allValues.length;
+
+							const allPubTypesSelected =
+								typeof value === "object" &&
+								value.pubTypes?.length === context.pubTypes.allValues.length;
+
+							if ((allStagesSelected || val.length === 0) && allPubTypesSelected) {
+								onChange(true);
+								return;
+							}
+
+							onChange({
+								stages:
+									val.length === 0
+										? context.stages.allValues
+										: (val as StagesId[]),
+								pubTypes:
+									typeof value === "object"
+										? value.pubTypes
+										: context.pubTypes.allValues,
+							});
+						}}
+						animation={0}
+						data-testid={`pub-${ApiAccessType.read}-stages-select`}
+					/>
+
+					<h3 className="font-semibold">Types</h3>
+					<span className="text-xs text-muted-foreground">
+						Select the types of Pubs this token can read
+					</span>
+					<MultiSelect
+						variant="inverted"
+						options={context.pubTypes.allOptions}
+						defaultValue={
+							value === true || !value ? context.pubTypes.allValues : value.pubTypes
+						}
+						onValueChange={(val) => {
+							const allPubTypesSelected =
+								val.length === context.pubTypes.allValues.length;
+
+							const allStagesSelected =
+								typeof value === "object" &&
+								value.stages?.length === context.stages.allValues.length;
+
+							if ((allStagesSelected || val.length === 0) && allPubTypesSelected) {
+								onChange(true);
+								return;
+							}
+
+							onChange({
+								pubTypes:
+									val.length === 0
+										? context.pubTypes.allValues
+										: (val as PubTypesId[]),
+								stages:
+									typeof value == "object"
+										? value.stages
+										: context.stages.allValues,
+							});
+						}}
+						animation={0}
+						data-testid={`pub-${ApiAccessType.read}-pubTypes-select`}
+					/>
+				</div>
+			);
+		},
```

Comment:
if we add another restriction at some point, we should make these components a bit DRYer